yjit/src/asm/x86_64/tests.rs

```diff
@@ -1,383 +1,280 @@
+        }
+        Ok(())
+    }
+}
 
-    ret(&mut cb);
+fn check_bytes<R>(bytes: &str, run: R) where R: FnOnce(&mut super::CodeBlock) {
+    let mut cb = super::CodeBlock::new();
+    run(&mut cb);
+    assert_eq!(format!("{:x}", cb), bytes);
+}
 
-    assert_eq!(0xC3, cb.mem_block[0]);
+#[test]
+fn test_add() {
+    check_bytes("80c103", |cb| add(cb, CL, imm_opnd(3)));
+    check_bytes("00d9", |cb| add(cb, CL, BL));
+    check_bytes("4000e1", |cb| add(cb, CL, SPL));
+    check_bytes("6601d9", |cb| add(cb, CX, BX));
+    check_bytes("4801d8", |cb| add(cb, RAX, RBX));
+    check_bytes("01d1", |cb| add(cb, ECX, EDX));
+    check_bytes("4c01f2", |cb| add(cb, RDX, R14));
+    check_bytes("480110", |cb| add(cb, mem_opnd(64, RAX, 0), RDX));
+    check_bytes("480310", |cb| add(cb, RDX, mem_opnd(64, RAX, 0)));
+    check_bytes("48035008", |cb| add(cb, RDX, mem_opnd(64, RAX, 8)));
+    check_bytes("480390ff000000", |cb| add(cb, RDX, mem_opnd(64, RAX, 255)));
+    check_bytes("4881407fff000000", |cb| add(cb, mem_opnd(64, RAX, 127), imm_opnd(255)));
+    check_bytes("0110", |cb| add(cb, mem_opnd(32, RAX, 0), EDX));
+    check_bytes("4883c408", |cb| add(cb, RSP, imm_opnd(8)));
+    check_bytes("83c108", |cb| add(cb, ECX, imm_opnd(8)));
+    check_bytes("81c1ff000000", |cb| add(cb, ECX, imm_opnd(255)));
 }
 
 #[test]
-fn test_push() {
-    let mut cb = CodeBlock::new();
+fn test_and() {
+    check_bytes("4421e5", |cb| and(cb, EBP, R12D));
+    check_bytes("48832008", |cb| and(cb, mem_opnd(64, RAX, 0), imm_opnd(0x08)));
+}
 
-    push(&mut cb, RAX);
+#[test]
+fn test_cmovcc() {
+    check_bytes("0f4ff7", |cb| cmovg(cb, ESI, EDI));
+    check_bytes("0f4f750c", |cb| cmovg(cb, ESI, mem_opnd(32, RBP, 12)));
+    check_bytes("0f4cc1", |cb| cmovl(cb, EAX, ECX));
+    check_bytes("480f4cdd", |cb| cmovl(cb, RBX, RBP));
+    check_bytes("0f4e742404", |cb| cmovle(cb, ESI, mem_opnd(32, RSP, 4)));
+}
 
-    assert_eq!(0x50, cb.mem_block[0]);
+#[test]
+fn test_cmp() {
+    check_bytes("38d1", |cb| cmp(cb, CL, DL));
+    check_bytes("39f9", |cb| cmp(cb, ECX, EDI));
+    check_bytes("493b1424", |cb| cmp(cb, RDX, mem_opnd(64, R12, 0)));
+    check_bytes("4883f802", |cb| cmp(cb, RAX, imm_opnd(2)));
 }
 
-/*
-// Print the bytes in a code block
-void print_bytes(codeblock_t* cb)
-{
-    for (uint32_t i = 0; i < cb->write_pos; ++i)
-    {
-        printf("%02X", (int)*cb_get_ptr(cb, i));
-    }
+#[test]
+fn test_cqo() {
+    check_bytes("4899", |cb| cqo(cb));
+}
 
-    printf("\n");
+#[test]
+fn test_jmp_with_rm_operand() {
+    check_bytes("41ffe4", |cb| jmp_rm(cb, R12));
 }
 
-// Check that the code block contains the given sequence of bytes
-void check_bytes(codeblock_t* cb, const char* bytes)
-{
-    printf("checking encoding: %s\n", bytes);
+#[test]
+fn test_lea() {
+    check_bytes("488d5108", |cb| lea(cb, RDX, mem_opnd(64, RCX, 8)));
+    // check_bytes("488d0500000000", |cb| lea(cb, RAX, mem_opnd(8, RIP, 0)));
```

Comment:
Should there be TODOs or something for the tests that are currently commented out?

Reply:
Added!

Reply:
I agree with Alan that the syntax of these looks quite nice. Well done.